src/meta_memcache/configuration.py

```diff
@@ -0,0 +1,137 @@
+import hashlib
+import socket
+from typing import Callable, NamedTuple
+
+from meta_memcache.base.connection_pool import ConnectionPool
+from meta_memcache.protocol import Key
+from meta_memcache.settings import DEFAULT_MARK_DOWN_PERIOD_S
+
+
+class IPPort(NamedTuple):
+    ip: str
+    port: int
+
+    def __str__(self) -> str:
+        return f"{self.ip}:{self.port}"
+
+
+def socket_factory_builder(
+    host: str,
+    port: int,
+    connection_timeout: float,
+    recv_timeout: float,
+    no_delay: bool,
+) -> Callable[[], socket.socket]:
+    """
+    Helper to generate a socket_builder with desired settings
+
+    The ConnectionPool class requires a callback to generate new
+    socket. This provides a helper to build it with the desired
+    configuration, but you can create your own and add TLS,
+    authentication, build unix sockets, etc.
+    """
+
+    def socket_builder() -> socket.socket:
+        s: socket.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        s.settimeout(connection_timeout)
+        s.connect((host, port))
+        if recv_timeout != connection_timeout:
+            s.settimeout(recv_timeout)
+        if no_delay:
```

Comment:
No need for keep alive options?

Reply:
TCP sockets are kept alive by default.

There is a lot of confusion about keep alive because of HTTP keepalives. HTTP 1.0 was closing conns after each request. Keepalive was added in 1.1 as application layer flag to keep the connection open.

On TCP, a socket is open forever by default. TCP keepalive is telling the tcp stack to do automatic probes on the connection from time to time. This is useful if the connection is not used frequently and you have a NAT that will timeout stablished tracked connections. Also on a stablished connection you don't know if the server is still alive on the other end until you send a pkt and resets or timeouts. For this case, memcache conns should be used very frequently and not via NAT or similar. I see no need for TCP keepalive. There is a noop meta command to do application level pings that can be implemented eventually and would work much better at ensuring the conn is functional than tcp keepalives.

Also the design of this library is offering users the option to pass their own socket builder to customize all the things they want. SSL, authentication, options... you name it. I want to stick to the basics and not support all possible user features :)

Reply:
As a note to this, we're using NAT on our side because we have khepri in the private network and unless we're connecting through the private network we're using a NAT to go outside

Reply:
What??? But the memcache is in the private net, right?

Reply:
Yes, luckily the memcached nodes are in private subnets hehehe